lib/web/server.cpp

```diff
@@ -0,0 +1,719 @@
+				sslKey = matcher[2];
+			}
+			else if(matcher[1] == "Redirect") {
+				host->redirect = true;
+				host->redirectHost = matcher[2];
+			}
+		}
+	}
+}
+
+bool Socialite::Web::Server::readHeader(NetStream* ns, HttpHeader* header, int* error) {
+	string lastLine = "-";
+	string line = ns->readLine(); // GET line
+	if(ns->isDead()) {
+		*error = -1;
+		return false;
+	}
+	header->cookies.erase(header->cookies.begin(), header->cookies.end()); // erase all cookies
+	header->contentLength = 0;
+	smatch matcher;
+	regex shape("^([A-Z]+) ([^\\s]+) HTTP/1.*$");
+
+	if(regex_match(line, matcher, shape)) {
+		regex_search(line, matcher, shape);
+		if(matcher[1] == "GET") {
+			header->type = GET;
+		}
+		else if(matcher[1] == "POST") {
+			header->type = POST;
+		}
+		else {
+			header->type = -1;
+		}
+
+		header->URI = matcher[2];
+	}
+	else {
+		cout << "Socialite::Web::Server::readHeader HEADER_BAD_REQUEST: '" << line << "'\n";
+		//ns->write(HEADER_BAD_REQUEST);
+		ns->close();
+		*error = 400;
+		return false;
+	}
+
+	// defaults
+	header->ifModifiedSince = (time_t) 0;
+
+	regex shape2("^([A-Za-z\\-]+): ([^]+)$");
+	while((line = ns->readLine()) != "" && lastLine != "") {
+		//cout << "Socialite::Web::Server::readHeader <<"<< line << ">>" << line.length() << ":" << (int) line[0] << "\n";
+		if(line == "\r") {
+			break; // ???? possibly a Chrome bug
+		}
+		lastLine = line;
+		if(regex_match(line, matcher, shape2)) {
+			regex_search(line, matcher, shape2);
+			if(matcher[1] == "Connection") {
+				if(matcher[2] == "keep-alive") {
+					header->closeType = KEEP_ALIVE;
+				}
+				else if(matcher[2] == "close") {
+					header->closeType = CLOSE;
+				}
+			}
+			else if(matcher[1] == "Cookie") {
+				//header->cookies = matcher[2];
+				string str = matcher[2];
+				smatch matcher2;
+				std::regex shapeCookie("([A-Za-z0-9\\-_]+)=([^;\r\n]*)");
+				while(std::regex_search(str, matcher2, shapeCookie)) {
+					header->cookies.insert(std::pair<string, string>(matcher2[1], string(matcher2[2])));
+					str = matcher2.suffix().str();
+				}
+			}
+			else if(matcher[1] == "Host") {
+				header->host = matcher[2];
+			}
+			else if(matcher[1] == "Content-Type") {
+				header->contentType = matcher[2];
+			}
+			else if(matcher[1] == "Content-Length") {
+				header->contentLength = std::stoi(matcher[2]);
+			}
+			else if(matcher[1] == "If-Modified-Since") {
+				header->ifModifiedSince = parseHttpTimestamp(matcher[2]);
+			}
+		}
+	}
+	//std::cout << "Socialite::Web::Server::readHeader empty line: " << line.length() << ": " << line << "\n";
+
+	if(header->type == POST && header->contentLength > 0) {
+		header->postData = ns->read(header->contentLength);
+		std::cout << header->postData << "\n";
+		mapPostData(header->postMap, header->postData);
+	}
+	//std::cout << "Socialite::Web::Server::readHeader done\n";
+	return true;
+}
+
+string Socialite::Web::Server::cgiLaunch(string file, HttpHeader* header, string getData) {
+	// int pipeFds[2];
```

Comment:
Need to find a way to put this back in... And it'd be great if we didn't have to call the custom `python_bridge` binary but rather if we could simply call the built in python interpreter that'd be pretty swell.